feat(website): add topbar and user dialog docs (#3249)
twilio-labs/paste: packages/paste-website/src/pages/components/topbar/index.mdx

  const feature = await getFeature('Topbar');
  return {
    props: {
      data: {
        ...packageJson,
        ...feature,
      },
      navigationData,
    },
  };
};

<NormalizedComponentHeader
  categoryRoute={SidebarCategoryRoutes.COMPONENTS}
  githubUrl="https://github.com/twilio-labs/paste/tree/main/packages/paste-core/components/topbar"
  storybookUrl="/?path=/story/components-topbar--"
  data={props.data}
/>

---

<contentwrapper>

<PageAside data={mdxHeadings} />

<content>

<LivePreview
  scope={{
    Topbar,
    Box,
    Input,
    Menu,
    MenuItem,
    useMenuState,
    MenuBadge,
    Badge,
    ProductSwitcher,
    ProductSwitcherButton,
    ProductSwitcherItem,
    useProductSwitcherState,
    LogoTwilioIcon,
    ProductSegmentIcon,
    ProductFlexIcon,
    ProductEmailAPIIcon,
    SearchIcon,
  }}
  noInline
>
  {defaultTopbarExample}
</LivePreview>

## Guidelines

### About Topbar

A Topbar is sticky to the top of the screen and consists of application-wide controls including:

- [Account Switcher](/components/account-switcher)
- Search [Input](/components/input)
- [User Dialog](#user-dialog)
- [Product Switcher](/components/product-switcher)

> ## Default Topbar

Reviewer: thought(non-blocking): Something about how there's a lot of repeated content between the "About Topbar" and "Default Topbar" sections bothers me, but I'll just wait til User Dialog is done and review the full page content.

Use a Topbar at the top of your application to house application-wide controls.

Account-specific controls are aligned to the left, such as:

- [Account Switcher](/components/account-switcher)
- Account type indicators, such as subaccount [Badges](/components/badge)
- Possible upgrade actions

User-specific controls are aligned to the right, such as:

- Search [Input](/components/input)
- Support [Button](/components/button)
- [Product Switcher](/components/product-switcher)
- [User Dialog](#user-dialog)

<LivePreview scope={{Topbar}} noInline>
  {emptyTopbarExample}
</LivePreview>

## User Dialog

> Coming soon!

Reviewer: Is User Dialog still going on this page?

Author: no

Author: it's its own package

  };
};

<NormalizedComponentHeader
  categoryRoute={SidebarCategoryRoutes.COMPONENTS}
  githubUrl="https://github.com/twilio-labs/paste/tree/main/packages/paste-core/components/topbar"
  storybookUrl="/?path=/story/components-topbar--"
  data={props.data}
/>

---

<contentwrapper>

<PageAside data={mdxHeadings} />

<content>

## Guidelines

### About Topbar

A Topbar is sticky to the top of the screen and includes application-wide controls including User Dialog (coming soon!), [Product Switcher](#product-switcher), [Account Switcher](#account-switcher), and search [Input](/components/input).

## Examples

### Default Topbar

Use a Topbar at the top of your application to house application-wide controls.

Account-specific controls are aligned to the left (e.g. [Account Switcher](#account-switcher), account type indicator [Badges](/components/badge), and possible upgrade actions).

User-specific controls are aligned to the right (e.g. search [Input](/components/input), support [Button](/components/button), [Product Switcher](#product-switcher), and User Dialog (coming soon!)).

<LivePreview scope={{Topbar}} noInline>
  {defaultTopbarExample}
</LivePreview>

## Account Switcher

> An Account Switcher is a stylized [Menu Badge](/components/menu#menu-badge) with a list of actions related to a user’s accounts.
> 
> Examples of account-related actions include navigating to other recently visited accounts, creating an account, and viewing all of a user’s accounts. Other names for accounts in Twilio products include workspaces and projects.
> 
> ### Accessibility
> 
> Refer to the accessibility section of [Menu](/components/menu).
> 
> ### Examples
> 
> Use an Account Switcher to display actions that a user can trigger related to their accounts.

Reviewer: ```suggestion
An Account Switcher is a stylized [Menu Badge](/components/menu#menu-badge) with a list of actions related to a user’s accounts. [Learn more about Account Switcher.](/components/account-switcher)
```

PR #3255 needs to get merged for the link to work, but we could take all the docs out here and redirect to the Account Switcher page.

<LivePreview
  scope={{
    useMenuState,
    Menu,
    MenuBadge,
    MenuItem,
  }}
  noInline
  showOverflow
>
  {defaultAccountSwitcherExample}

> </LivePreview>

Reviewer: chore: remove this since Account Switcher has a dedicated docs page.

## User Dialog

Coming soon!

## Product Switcher

> A Product Switcher is an opinionated menu that presents a list of Twilio products that a user has access to and can switch between.
> 
> ### Accessibility
> 
> Refer to the accessibility section of [Menu](/components/menu).
> 
> ### Examples
> 
> Use a Product Switcher for when a user has access to multiple Twilio products so a user can switch between them.
> 
> The default state of a Product Switcher item indicates that the control is static or not selected. A Product Switcher item can be placed into a selected state in the same way as a radio button. Use the `checked` prop to set a selected item by default.

Reviewer: ```suggestion
A Product Switcher is an implementation of [Menu](/components/menu) that shows the Twilio products a user has access to and can switch between. [Learn more about Product Switcher.](/components/product-switcher)
```

<LivePreview
  scope={{
    ProductSwitcher,
    ProductSwitcherButton,
    ProductSwitcherItem,
    useProductSwitcherState,
    LogoTwilioIcon,
    ProductSegmentIcon,
    ProductFlexIcon,
    ProductEmailAPIIcon,
  }}
  noInline
  showOverflow
>
  {defaultProductSwitcherExample}

> </LivePreview>

Reviewer: todo: replace with the example from PR #3255

> </LivePreview>

Reviewer: todo: swap this out for the Product Switcher example in PR https://github.com/twilio-labs/paste/pull/3255

## Usage guide

### Figma

> Topbar, Account Switcher, Product Switcher, and User Dialog components are available in the Paste Components library.

Reviewer: ```suggestion
Topbar, Account Switcher, Product Switcher, and User Dialog components are available in the Paste Components library as [Navigation components](https://www.figma.com/file/l2qtpQVMmXzQPVB3rEAA8i/%5BNEW%5D-Paste-Components?type=design&node-id=27837-187276&t=xcPacI1AN13MooNC-0).
```

### API

#### Installation

```bash
yarn add @twilio-paste/topbar - or - yarn add @twilio-paste/core
yarn add @twilio-paste/menu - or - yarn add @twilio-paste/core
yarn add @twilio-paste/product-switcher - or - yarn add @twilio-paste/core
```

#### Usage

```jsx
import {Topbar} from '@twilio-paste/core/topbar';

const Topbar = () => {
  return (
    <Topbar>
      {Topbar contents}
    </Topbar>
  );
};
```

#### Topbar props

| Prop      | Type              | Description                                                                               | Default  |
| --------- | ----------------- | ----------------------------------------------------------------------------------------- | -------- |
| children? | `React.ReactNode` | Recommended: Account Switcher, User Dialog, Product Switcher, Search Input, etc.          |          |
| element?  | `string`          | Overrides the default element name to apply unique styles with the Customization Provider | 'TOPBAR' |

#### Account Switcher props

Refer to [Menu props](/components/menu#props)

#### Product Switcher props

> useProductSwitcher props extends [useMenuState props](/components/menu#usemenustate)
> 
> ProductSwitcherProps extends [MenuProps](/components/menu#menu)
> 
> ProductSwitcherButtonProps extends [MenuButtonProps](/components/menu#menubutton)
> 
> ProductSwitcherItemProps extends [MenuItemRadioProps](/components/menu#menuitemradio) plus the following properties:

Reviewer: ```suggestion
useProductSwitcher props extends [useMenuState props](/components/menu#usemenustate).

ProductSwitcherProps extends [MenuProps](/components/menu#menu).

ProductSwitcherButtonProps extends [MenuButtonProps](/components/menu#menubutton).

ProductSwitcherItemProps extends [MenuItemRadioProps](/components/menu#menuitemradio) plus the following properties:
```

| Prop             | Type                           | Description | Default |
| ---------------- | ------------------------------ | ----------- | ------- |
| productName      | `string`                       |             |         |
| productStrapline | `string`                       |             |         |

> | productIcon      | `NonNullable<React.ReactNode>` |             |         |

Reviewer: chore: remove
